packages/core/custom-branding/core-custom-branding-server-internal/custom_branding_service.ts

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+import type { CustomBrandingFetchFn, CustomBrandingStart } from '@kbn/core-custom-branding-server';
+import type { CustomBranding } from '@kbn/core-custom-branding-common';
+import { KibanaRequest } from '@kbn/core-http-server';
+import type { CoreContext } from '@kbn/core-base-server-internal';
+import type { Logger } from '@kbn/logging';
+
+/**
+ * @internal
+ */
+export interface InternalCustomBrandingSetup {
+  register: (pluginName: string, fetchFn: CustomBrandingFetchFn) => void;
+  getBrandingFor: (request: KibanaRequest) => Promise<CustomBranding>;
+}
+
+export class CustomBrandingService {
+  private pluginName: string | undefined;
+  private logger: Logger;
+  private fetchFn: CustomBrandingFetchFn | undefined;
+  private startCalled: boolean = false;
+
+  constructor(coreContext: CoreContext) {
+    this.logger = coreContext.logger.get('custom-branding-service');
+  }
+
+  public setup(): InternalCustomBrandingSetup {
+    return {
+      register: (pluginName, fetchFn) => {
+        this.logger.info('CustomBrandingService registering plugin: ' + pluginName);
+        if (this.pluginName) {
+          throw new Error('Another plugin already registered');
+        }
+        this.pluginName = pluginName;
+        this.fetchFn = fetchFn;
+      },
+      getBrandingFor: this.getBrandingFor,
+    };
+  }
+
+  public start(): CustomBrandingStart {
+    this.startCalled = true;
+    return {};
+  }
+
+  public stop() {}
+
+  private getBrandingFor = async (request: KibanaRequest): Promise<CustomBranding> => {
+    if (!this.startCalled) {
+      throw new Error('Cannot be called before #start');
+    }
+    if (!this.pluginName || this.pluginName !== 'customBranding') {
+      return {};
+    }
+    return await this.fetchFn!(request);
```

Comment:
This is 'dangerous', as it can throw if no plugin registered the `fetchFn`. 

I would replace to something like

```ts
return this.fetchFn ? await this.fetchFn!(request) : someEmptyDefaultBranding;

```

Reply:
This scenario would ideally be caught by a previous line; if no `pluginName` there is no `fetchFn` either. It shouldn't be possible to register a plugin without providing a `fetchFn`. I'm more in favor of changing `register` function then that would explicitly throw if `fetchFn` is not provided as an argument. Wdyt?